### PR TITLE
Full review of all existing code

### PR DESCRIPTION
**Pull request by ryanholmdahl**
- [ ] Make `DagLayout` functional
    - [ ] Test Python bindings
    - [x] Implement JS bindings
    - [ ] Add interactions to component
- [x] Test Javascript bindings

### CODE REVIEW

File: bindings/python/vizstack/assembler.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+from vizstack.types import JsonType
+from vizstack.view import get_view, View
+
+
+__all__ = ['assemble']
+
+
+def assemble(obj: Any) -> Dict[str, Union[str, Dict[str, Dict[str, JsonType]]]]:
+    """Returns a string which contains all of the information needed to render a visualization of `obj`.
```
Comment (nikhilxb) [inline bindings/python/vizstack/assembler.py]: Update

File: bindings/python/vizstack/assembler.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            'rootId': the View ID of the top-level View in 'models'.
+        }
+    """
+    to_add: List[View] = [get_view(obj)]
+    added: MutableSet[str] = set()
+    return_dict: Dict[str, Union[Optional[str], Dict[str, JsonType]]] = {
+        'rootId': to_add[0].id,
+        'models': dict(),
+    }
+    while len(to_add) > 0:
```
Comment (nikhilxb) [inline bindings/python/vizstack/assembler.py]: Implementation comment

File: bindings/python/vizstack/view.py [hunk 3]
```diff
@@ -0,0 +1,784 @@
...
+        A View which describes how to render ``o``.
+    """
+    global _CURRENT_PLACEHOLDERS
+    if id(o) in _CURRENT_PLACEHOLDERS:
+        return _CURRENT_PLACEHOLDERS[id(o)]
+    _CURRENT_PLACEHOLDERS[id(o)] = View()
+
+    is_switch = False
+
+    if isinstance(o, View):
```
Comment (nikhilxb) [inline bindings/python/vizstack/view.py]: Add comments

File: bindings/python/vizstack/view.py [hunk 4]
```diff
@@ -0,0 +1,784 @@
...
+# this happens, `id(a)` will already by in `_CURRENT_PLACEHOLDERS`, so instead of trying to generate a new `View` for
+# `a` again, we immediately return `_CURRENT_PLACEHOLDERS[id(a)]`. This breaks the infinite loop, and the first call to
+# `get_view(a)` is able to finish creating the `View` object for `a`. Before the function returns, the generic `View`
+# at `_CURRENT_PLACEHOLDERS[id(a)]` will be mutated to be identical to the `View` that was generated for `a`,
+# and is then removed from `_CURRENT_PLACEHOLDERS`. By the end of any top-level call to `get_view()`,
+# `_CURRENT_PLACEHOLDERS` will be empty.
+_CURRENT_PLACEHOLDERS = dict()
+
+
+def get_view(o: Any) -> Union['View']:
```
Comment (ryanholmdahl) [inline bindings/python/vizstack/view.py]: No need for `Union`

File: bindings/python/vizstack/view.py [hunk 5]
```diff
@@ -0,0 +1,784 @@
...
+# At the start of the first call to `get_view(a)`, we create an empty generic `View` instance and set
+# `_CURRENT_PLACEHOLDERS[id(a)] = View()`. The placeholder is an empty object at this point. The process
+# then continues as in the naive implementation, and eventually `get_view(a)` is called for the second time. When
+# this happens, `id(a)` will already by in `_CURRENT_PLACEHOLDERS`, so instead of trying to generate a new `View` for
+# `a` again, we immediately return `_CURRENT_PLACEHOLDERS[id(a)]`. This breaks the infinite loop, and the first call to
+# `get_view(a)` is able to finish creating the `View` object for `a`. Before the function returns, the generic `View`
+# at `_CURRENT_PLACEHOLDERS[id(a)]` will be mutated to be identical to the `View` that was generated for `a`,
+# and is then removed from `_CURRENT_PLACEHOLDERS`. By the end of any top-level call to `get_view()`,
+# `_CURRENT_PLACEHOLDERS` will be empty.
+_CURRENT_PLACEHOLDERS = dict()
```
Comment (nikhilxb) [inline bindings/python/vizstack/view.py]: Lower case convention

File: bindings/python/vizstack/view.py [hunk 6]
```diff
@@ -0,0 +1,784 @@
...
+        is_switch = True
+    elif isinstance(o, tuple):
+        view = _SwitchSequence(
+            list(o),
+            start_motif='Tuple[{}] ('.format(len(o)),
+            end_motif=')',
+            summary='Tuple[{}]'.format(len(o)))
+        is_switch = True
+    elif isinstance(o, dict):
+        view = _SwitchKeyValues(
```
Comment (nikhilxb) [inline bindings/python/vizstack/view.py]: `_SwitchKeyValue`

File: bindings/python/vizstack/view.py [hunk 7]
```diff
@@ -0,0 +1,784 @@
...
+                continue
+        view = _SwitchKeyValues(
+            contents,
+            item_separator='=',
+            start_motif='Instance[{}] {{'.format(type(o).__name__),
+            end_motif='}',
+            summary='Instance[{}]'.format(type(o).__name__))
+        is_switch = True
+    if is_switch:
+        _CURRENT_PLACEHOLDERS[id(o)].__mutate__(Switch(view._modes[-1:] + view._modes[:-1], view._items))
```
Comment (nikhilxb) [inline bindings/python/vizstack/view.py]: Document this non trivial logic

File: bindings/python/vizstack/view.py [hunk 8]
```diff
@@ -0,0 +1,784 @@
...
+        return _CURRENT_PLACEHOLDERS[id(o)]
+    _CURRENT_PLACEHOLDERS[id(o)] = View()
+
+    is_switch = False
+
+    if isinstance(o, View):
+        view = o
+    elif hasattr(o, '__view__'):
+        view = getattr(o, '__view__')()
+    elif isinstance(o, (str, int, float, bool)) or o is None:
```
Comment (nikhilxb) [inline bindings/python/vizstack/view.py]: Refactor into isolated language_defaults function

File: bindings/python/vizstack/view.py [hunk 9]
```diff
@@ -0,0 +1,784 @@
...
+# objects.
+#
+# Subtle stuff:
+#  - Every "slot" (e.g., a `Switch` mode or a `DagLayout` cell) must have an item attached to it, but not every item
+#    needs an existing mode; such items will not be included in the assembled `View`.
+# ======================================================================================================================
+
+class View:
+    def __init__(self):
+        self.id: str = '@id:{}'.format(cuid.cuid())
```
Comment (nikhilxb) [inline bindings/python/vizstack/view.py]: Comment

File: bindings/python/vizstack/view.py [hunk 10]
```diff
@@ -0,0 +1,784 @@
...
+        self._meta: Dict[str, JsonType] = {}
```
Comment (nikhilxb) [inline bindings/python/vizstack/view.py]: Comment

File: bindings/python/vizstack/view.py [hunk 11]
```diff
@@ -0,0 +1,784 @@
...
+        Args:
+            key:
+            value:
+
+        Returns:
+
+        """
+        self._meta[key] = value
+
+    def __mutate__(self, view: 'View') -> None:
```
Comment (nikhilxb) [inline bindings/python/vizstack/view.py]: Change from dunder

Comment (nikhilxb) [reply bindings/python/vizstack/view.py]: TODO: Think about with Nikhil

File: bindings/python/vizstack/view.py [hunk 12]
```diff
@@ -0,0 +1,784 @@
...
+            grid.cell('{}'.format(i), *current_position, 1, 1)
+            grid.item(item, '{}'.format(i))
+            current_position[update_index] += 1
+        if self._end_motif is not None:
+            grid.cell('end_motif', *current_position, 1, 1)
+            grid.item(self._end_motif, 'end_motif')
+        return grid.assemble()
+
+
+class KeyValues(View):
```
Comment (nikhilxb) [inline bindings/python/vizstack/view.py]: KeyValue

File: bindings/python/vizstack/view.py [hunk 13]
```diff
@@ -0,0 +1,784 @@
...
+                        current_char = c
+                    cell_bounds[current_char]['X'] = row_len
+            for cell_name, cell_bound in cell_bounds.items():
+                self._cells[cell_name] = {
+                    'col': cell_bound['x'],
+                    'row': cell_bound['y'],
+                    'width': cell_bound['X'] - cell_bound['x'],
+                    'height': cell_bound['Y'] - cell_bound['y'],
+                }
+        # TODO: assert non-overlapping
```
Comment (nikhilxb) [inline bindings/python/vizstack/view.py]: Does CSS render on top

File: components/src/interaction/manager.js [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+type ViewerInfo = {|
+    id: ViewerId,
+    parent?: ViewerHandle,
+|};
+
+/** Fields in a `ViewerHandle` which are dependent upon the type of `ViewModel` being rendered by the `Viewer`. */
+export type ComponentHandle = TextPrimitiveHandle | ImagePrimitiveHandle | SwitchLayoutHandle | GridLayoutHandle | FlowLayoutHandle;
+
+/** Provides information about a `Viewer`, as well as methods which alter its state. */
+export type ViewerHandle = {|
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: Nikhil: Separate into individual for Typescript

File: components/src/interaction/manager.js [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+    };
+
+    /**
+     * Adds a new handler for a given topic.
+     *
+     * @param topic
+     * @param handler
+     */
+    on<E: Event>(topic: $PropertyType<E, 'topic'>,
+                 handler: EventHandler<$PropertyType<E, 'message'>>) {
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: Need parameter bc want event handler to have type of msg.

File: components/src/interaction/manager.js [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+        return new InteractionSet(this.viewers.filter((viewer) => fn(viewer)));
+    }
+
+    forEach(fn: (viewer: ViewerHandle) => void): void {
+        this.viewers.forEach((viewer) => fn(viewer));
+    }
+}
+
+/** The type of function which is called when an `Event` is emitted to an `InteractionManager`. */
+type EventHandler<Message> = (all: InteractionSet, message: Message, global: any) => void
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: Consider renaming to `MessageHandler`

File: components/src/interaction/manager.js [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+
+    /**
+     * Enqueues a new `Event` to be processed. If no `Event` is currently being processed, then that
+     * `Event` is processed immediately.
+     *
+     * @param topic
+     * @param message
+     */
+    emit: <E: Event>(topic: $PropertyType<E, 'topic'>,
+                     message: $PropertyType<E, 'message'>) => void =
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: @ryanholmdahl: Implementation comment about why this syntax.

File: components/src/interaction/manager.js [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+
+    /** Dequeues an `Event` from `this.eventQueue` and fires all handler functions for that
+     * `Event`. */
+    _processNextEvent() {
+        this.processingEvent = true;
+        const { topic, message } = this.eventQueue.shift();
+        if (topic in this.handlers) {
+            this.handlers[topic].forEach(
+                (handler) => handler(
+                    // Flow doesn't know how to handle `Object.values()`, so we have to cast it to
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: Moving to Typescript

File: components/src/interaction/manager.js [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+                    // `any` then cast the `handleFactory` to its proper type
+                    new InteractionSet((Object.values(this.viewers): any).map((handleFactory: () => ViewerHandle) => handleFactory())),
+                    message,
+                    this.global,
+                )
+            );
+        }
+        this.processingEvent = false;
+        if (this.eventQueue.length > 0) {
+            this._processNextEvent();
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: Rewrite as while loop

File: components/src/interaction/manager.js [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+import type {ViewerDidMouseOverEvent, Event, ViewerDidMouseOutEvent, ViewerDidClickEvent } from './events';
+import type {TextPrimitiveHandle} from '../primitives/TextPrimitive';
+import * as React from "react";
+import type {ImagePrimitiveHandle} from "../primitives/ImagePrimitive";
+import type {SwitchLayoutHandle} from "../layouts/SwitchLayout";
+import type {GridLayoutHandle} from "../layouts/GridLayout";
+import type {FlowLayoutHandle} from "../layouts/FlowLayout";
+
+export type ViewerId = string;
+
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: @ryanholmdahl : Want high level documentation of flow.

File: components/src/interaction/manager.js [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+        if (useMouseDefaults) {
+            this._useMouseDefaults();
+        }
+        if (useKeyboardDefaults) {
+            this._useKeyboardDefaults(documentElement);
+        }
+    }
+
+    _useMouseDefaults() {
+        this.on<ViewerDidMouseOverEvent>('Viewer.DidMouseOver',
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: Before all `this.on` want 1-liner comment about what the behavior is.

File: components/src/interaction/manager.js [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+    id(ids: ViewerId | Array<ViewerId>): InteractionSet {
+        if (Array.isArray(ids)) {
+            return this.filter((viewer) => ids.includes(viewer.id));
+        }
+        else {
+            return this.filter((viewer) => viewer.id === ids);
+        }
+    }
+
+    type(types: $PropertyType<ViewModel, 'type'> | Array<$PropertyType<ViewModel, 'type'>>) {
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: Better typechecking for `InteractionSet<>`

File: components/src/interaction/manager.js [hunk 23]
```diff
@@ -0,0 +1,314 @@
...
+    // The global state which is accessed by `EventHandler` functions.
+    global: any = {};
+    // The queue of `Event`s which have been emitted but whose handlers have not yet been executed.
+    eventQueue: Array<Event> = [];
+    // Whether the `InteractionManager` is currently executing the handlers for an event.
+    processingEvent: boolean = false;
+
+    /**
+     *
+     * @param options: `documentElement` is needed, since `document` might not be correct in all
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: This is bad

File: components/src/interaction/manager.js [hunk 24]
```diff
@@ -0,0 +1,314 @@
...
+                all.id(message.viewerId).forEach((viewer) => {
+                    if (viewer.model.type === 'SwitchLayout') {
+                        viewer.doIncrementMode();
+                    }
+                });
+            }
+        );
+    }
+
+    _useKeyboardDefaults(documentElement: Document | HTMLElement) {
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: Bad documentElement

File: components/src/interaction/manager.js [hunk 25]
```diff
@@ -0,0 +1,314 @@
...
+
+        this.on<{|topic: 'KeyDown', message: {|key: string|}|}>('KeyDown',
+            (all, message, global) => {
+                all.id(global.selected).forEach((viewer) => {
+                    if (message.key === 'Enter') {
+                        console.log('processed');
+                        if (viewer.selectedViewerId) {
+                            global.selected = viewer.selectedViewerId;
+                            viewer.doUnhighlight();
+                            all.id(global.selected).forEach((viewer) => viewer.doHighlight());
```
Comment (nikhilxb) [inline components/src/interaction/manager.js]: Need to separate "highlight" (mouse, other) from "cursor" (keyboard).

**Comment by nikhilxb** [comment]
@ryanholmdahl What is the status of this PR? What do we need to do to finally merge this branch?
[] Convert @vizstack/viewer to Typescript

**Comment by nikhilxb** [comment]
Looks great!!!!!!!!

**Pull request merged**